ipc/provider/src/checkpoint.rs

```diff
@@ -162,30 +178,71 @@ impl<T: BottomUpCheckpointRelayer + Send + Sync + 'static> BottomUpCheckpointMan
                 log::debug!("bottom up bundle: {bundle:?}");
 
-                let epoch = self
-                    .parent_handler
-                    .submit_checkpoint(
-                        submitter,
-                        bundle.checkpoint,
-                        bundle.signatures,
-                        bundle.signatories,
-                    )
+                // We support parallel checkpoint submission using FIFO order with a limited parallelism (controlled by
+                // the size of submission_semaphore).
+                // We need to acquire a permit (from a limited permit pool) before submitting a checkpoint.
+                // We may wait here until a permit is available.
+                let parent_handler_clone = Arc::clone(&self.parent_handler);
+                let submission_permit = self
+                    .submission_semaphore
+                    .clone()
+                    .acquire_owned()
                     .await
-                    .map_err(|e| {
-                        anyhow!(
-                            "cannot submit bottom up checkpoint at height {} due to: {e:}",
-                            event.height
-                        )
-                    })?;
-
-                log::info!(
-                    "submitted bottom up checkpoint({}) in parent at height {}",
-                    event.height,
-                    epoch
+                    .unwrap();
+                all_submit_tasks.push(tokio::task::spawn(async move {
+                    match Self::submit_checkpoint(parent_handler_clone, submitter, bundle, event)
+                        .await
+                    {
+                        Ok(()) => {
+                            log::debug!("Successfully submitted checkpoint");
+                        }
+                        Err(_err) => {
+                            log::debug!("Failed to submit checkpoint");
+                        }
+                    }
+                    drop(submission_permit);
+                }));
+
+                count += 1;
+                log::debug!(
+                    "This round has asynchronously submitted {:} checkpoints",
+                    count
                 );
             }
         }
 
+        log::debug!("Waiting for all submissions to finish");
+        join_all(all_submit_tasks).await;
+
+        Ok(())
+    }
+
+    async fn submit_checkpoint(
+        parent_handler: Arc<T>,
+        submitter: Address,
+        bundle: BottomUpCheckpointBundle,
+        event: QuorumReachedEvent,
+    ) -> Result<(), anyhow::Error> {
+        let epoch = parent_handler
+            .submit_checkpoint(
+                &submitter,
+                bundle.checkpoint,
+                bundle.signatures,
+                bundle.signatories,
+            )
+            .await
+            .map_err(|e| {
+                anyhow!(
+                    "cannot submit bottom up checkpoint at height {} due to: {e:}",
```

Comment:
```suggestion
                    "cannot submit bottom up checkpoint at height {} due to: {e:?}",
```
Not sure what `{e:}` is but I suspect it's the same as `{e}`. There is also `{e:#}`.